framework/include/base/Adaptivity.h

```diff
@@ -129,9 +129,11 @@ class Adaptivity : public ConsoleStreamInterface,
   void setRecomputeMarkersFlag(const bool flag) { _recompute_markers_during_cycles = flag; }
 
   /**
-   * Switch from h-refinement to p-refinement
+   * Indicate whether the kind of adaptivity we're doing is p-refinement
+   * @param doing_p_refinement Whether we're doing p-refinement
+   * @param disable_lagrange_p_refinement Whether to disable p-refinement of Lagrange variables
    */
-  void switchHToPRefinement();
+  void doingPRefinement(bool doing_p_refinement, bool disable_lagrange_p_refinement);
```

Comment:
what s the hate against Lagrange p-refinement

Reply:
ok it s against nodal

@roystgnr am I correct in that the problem (needing a second order mesh to do p-refinement) is the same for:
HERMITE,  CLOUGH in general (some dofs at nodes, some gradient dofs)
HIERARCHIC, SZABAB, BERNSTEIN, RATIONAL_BERNSTEIN at some orders have dofs on nodes

Reply:
It's not directly a matter of p refinement, it's a matter of the combination of p with FEFamily.  As soon as you get a DoF that has support on a set of elements, however you got that DoF, you need a node which is connected to precisely that set of elements.  And even a second order mesh may not be enough - a TET10 will do p=1 or p=2 fine, but get up to p=3 on a C0 element and now you've got face degrees of freedom and you need a TET14 to give you face nodes to put them on.  That's true whether you got p=3 by declaring it from the get-go or by p-refining a p=2 variable once or a p=1 variable twice.  Every C0 or C1 variable type will get to the point where it needs edge and face nodes at *some* degree, depending on the element type; only the completely discontinuous bases don't ever need them.

Lagrange is extra tricky because since it's nodal rather than hierarchic we can't just get away with having e.g. one face node, we have to have one face node at exactly the spots where we have one Lagrange DoF evaluation for that p.  So we can support p=1 Lagrange on any element or p=2 on a quadratic (or in the future a quartic or any p=2*n element), but we *can't* support p=2 Lagrange on a cubic element.  A Hex64 will have 4 face nodes, and will work fine with a cubic Lagrange, but none of those nodes are in the center of the face where quadratic Lagrange needs a node.  So unless you "skip" p levels, p refinement on Lagrange can go from 1 to 2 and then it gets stuck and that's just sad.

Rational_Bernstein (the reason why I'll be adding Hex64 and other higher order Hex/Quad/Edge elements)  is kind of an in-between case - the Bernstein DoFs aren't point evaluations on nodes, but they're associated with nodes and the Rational<Foo> weightings get stored on nodes.  IGA meshes will have some of the same restrictions as Lagrange for that reason.

Reply:
so it's a matter of which family uses the additional second / third etc order nodes to start dofs. 
Just Lagrange and Rational Bernstein?

Reply:
`LAGRANGE_VEC` is also nodal (albeit multiple components per node).  I think that's the only other that can't be made to work easily with uniform p refinement.

With *adaptive* p refinement, though, we have the restriction that libMesh currently only constructs constraints for hierarchic bases (where the p=n+1 basis is just a superset of the p=n basis).  Discontinuous FE don't need constraints, but Bernstein and Clough need constraints that we can't construct for them.

Oh, and there's the fact that not everything that can support higher p in theory has the code written in practice.  `FEInterface::max_order` lists a bunch of not-implemented cases still: Szabo-Babuska is only hard-coded up to p=7, Clough for p=3, Nedelec-One and Raviart-Thomas for p=1.  We could add higher order capability to any of those but nobody has.

Reply:
ok for Lagrange_vec being also purely nodal.

So both Szabab and Clough have DOFs at nodes and max allowed degrees that can be considered for uniform or adaptive p-refinement. But are they using the edge nodes of a TET10 and/or the face nodes of a TET14 in a way that precludes from hosting a third order on a TET10 / a second order on a TET14  ?

> It's not directly a matter of p refinement, it's a matter of the combination of p with FEFamily. 

this is what I'm trying to get to. Is it really just Lagrange variables that are problematic, or any variable in which we store DOFs at nodes will be problematic And if the latter, should Alex just replace lagrange to 'NodalDofed' to make it work in the text

Reply:
I could make this a parameter for which users can supply a list of families to disable, or variable groups but I think that families would be more user friendly. The only thing I really care about is getting this into users' hands sooner than later.

Reply:
> or any variable in which we store DOFs at nodes will be problematic

Any variable in which we store DoFs at nodes *can* be problematic, but doesn't have to be on the right mesh.  Do `all_complete_order()` (or load a mesh from file where all the Tets are TET14 etc., whatever), and anything for which `FE::is_hierarchic()` will be fine to p-refine up to its max order.

Reply:
I could change this to `disable_non_hierarchic`?

Reply:
That seems like a solid improvement to me.

Looking in libMesh, all our safety checking is in devel mode (because it necessarily involves a test at every element/neighbor interface with a p mismatch), and that's not good enough for MOOSE users, so better to disable p-refinement of non-hierarchic elements entirely.  We could selectively re-enable it if people really want to use one of them for supported *uniform* p refinement.

Reply:
I think Roy is saying this is only OK for all_complete_order meshes.

so it s ok if you assert after that the mesh is as expected?

Reply:
> so it s ok if you assert after that the mesh is as expected?

I don't know what "that" means, or what I'm supposed to be asserting.

It's certainly valid for a user running with DFEM with monomial variables (a large user base) to run with a mesh that is not `all_complete_order()`

Reply:
well monomial is fine for this problem of nodal values right
I think the concern is has_nodal_dofs_non_hierarchic variables

Reply:
FWIW we'll already be hitting decent error messages if we overrefine an FE without all the right Elem types to support it; the same one we hit now if someone tries to use e.g. second order Lagrange on a HEX8.

Reply:
Though I admit an error before refining would be more user friendly than an error after.

Reply:
@GiudGiud sure, but I don't understand what you're proposing we do about it. Re-implement the exact same error messages as already exist in libMesh but just do it before we p-refine? I don't see how much we gain there. Generally, we cannot know how much p-refinement there will be *a priori* at the start of a simulation, so I don't know how much user time you will save by giving an error message right before p-refinement as opposed to right after. 

I think what does make sense is to allow the user to disable p-refinement of different finite element families. Perhaps as I suggested before this parameter can be changed to be a `MultiMooseEnum` and the user can specify what families they want to disable because they know that they might or will hit a p-refinement error and they don't care that a given variable have the p-order applied to it. Or if we really want to grant flexibility it can be a `std::vector<MultiMooseEnum>` in which the user can specify disabling of p-refinement for FEFamilies of given systems.

Reply:
Ok I see your point. If the errors already exist in libmesh, we're already covered.
It seems fitting to make it a MultiMooseEnum, so we catch both lagrange and lagrange vec from the get go